### DISCUSSION TITLE
Can greenfield have Web Applications that show web content directly?

### DISCUSSION THREAD

**Discussion by fpemud** [udevbe/greenfield#38] [Ideas]
> A Web application must render it’s content to an ArrayBuffer as Web Workers do not have access to the DOM and as such can not use HTML to render their content

Is it good to have these 2 extra kinds of "web application" in greenfield:
1. Shows web element directly. Like this: [https://github.com/os-js/OS.js](url)
    This kind of web application would be easier to write and more effecient than rendering bitmap content into ArrayBuffer.
2. A http server is started by application end-point server when app runs,  and the app code run in browser using <iframe src="..."> to show the web content from the newly started http server.
    This kind of web application would be more effecient than transfering h264 stream from server to browser.
    But this kind of web application needs extra protocol between browser and application end-point server.

**Comment by Zubnix** [comment]
I've been considering this as well. They way it currently works is that each application is shown as a webgl texture (mostly for speed an atomic updating of the screen). However the problem with the iframe approach is that [it cannot be used as a webgl texture](https://bugs.chromium.org/p/chromium/issues/detail?id=605714) or [at least not very efficiently](https://surma.dev/things/dom2texture/):

> This approach is not cheap. At all. For the last demo, going from SVG to rendered canvas takes 10ms on the new 2017 MacBook Pro, and 62ms on a Pixel. Considering the frame budget for 60fps is 16ms, you need to be extremely careful.

[However a lot of other people seems to want this as well](https://medium.com/webmr/dom-to-canvas-for-webxr-in-just-7-years-e032060437fb), so maybe one day it will be possible.

Another option is to hide/move aside the webgl canvas that draws all other applications (ie your workspace) and replace it with the application's iframe and present it to the user as an always on-top fullscreen application, but in no way can you show both the iframe app and another app in an overlapping floating window style.

This is afaik the only hard constraint, everything else (ie presenting the iframe app as a Wayland appliction) should be solvable.
